source/Calamari/Kubernetes/Conventions/HelmUpgradeWithKOSConvention.cs

```diff
@@ -0,0 +1,554 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Calamari.Common.Commands;
+using Calamari.Common.Features.Packages;
+using Calamari.Common.Features.Processes;
+using Calamari.Common.Features.Scripts;
+using Calamari.Common.FeatureToggles;
+using Calamari.Common.Plumbing.FileSystem;
+using Calamari.Common.Plumbing.Logging;
+using Calamari.Common.Plumbing.Variables;
+using Calamari.Deployment.Conventions;
+using Calamari.Kubernetes.Helm;
+using Calamari.Kubernetes.Integration;
+using Calamari.Kubernetes.ResourceStatus;
+using Calamari.Kubernetes.ResourceStatus.Resources;
+using Calamari.Util;
+using Newtonsoft.Json;
+using Octopus.CoreUtilities.Extensions;
+using YamlDotNet.RepresentationModel;
+
+namespace Calamari.Kubernetes.Conventions
+{
+    public class HelmUpgradeWithKOSConvention : IInstallConvention
+    {
+        readonly ILog log;
+        readonly ICommandLineRunner commandLineRunner;
+        readonly ICalamariFileSystem fileSystem;
+        readonly HelmTemplateValueSourcesParser valueSourcesParser;
+        readonly IResourceStatusReportExecutor statusReporter;
+        readonly IManifestReporter manifestReporter;
+        readonly Kubectl kubectl;
+
+        public HelmUpgradeWithKOSConvention(ILog log,
```

Comment:
This has become a bit of a mamoth class. Could we break it up into orchestrating the two concerns, execution vs polling\parsing?
It doesnt _look_ like there should be shared state between them, but it was hard to tell :D

Reply:
I should be able to yeah

Reply:
Addressed in [2889b8a](https://github.com/OctopusDeploy/Calamari/pull/1374/commits/2889b8a9e0435b4187f109d158ebe50d7cc9b191) and [14025aa](https://github.com/OctopusDeploy/Calamari/pull/1374/commits/14025aae75ca84cd4884e4c7d7e85eff9ad99ea1)